packages/apollo-server-core/src/requestPipeline.ts

```diff
@@ -162,35 +164,64 @@ export async function processGraphQLRequest<TContext>(
     requestContext,
   });
 
-  const parsingDidEnd = await dispatcher.invokeDidStartHook(
-    'parsingDidStart',
-    requestContext,
-  );
-
   try {
-    let document: DocumentNode;
-    try {
-      document = parse(query);
-      parsingDidEnd();
-    } catch (syntaxError) {
-      parsingDidEnd(syntaxError);
-      return sendErrorResponse(syntaxError, SyntaxError);
+    // If we're configured with a document store, we'll utilize the hash of the
+    // operation in order to lookup a previously parsed-and-validated operation
+    // from that.  A failure to retrieve anything from the cache simply means
+    // we're guaranteed to do the parsing and validation ourselves.
+    let document: DocumentNode | undefined;
+
+    if (config.documentStore) {
+      document = await config.documentStore.get(queryHash);
     }
 
-    requestContext.document = document;
+    // If we still don't have a document, we'll need to parse and validate it.
+    // With success, we'll attempt to save it into the store for future use.
+    if (!document) {
+      const parsingDidEnd = await dispatcher.invokeDidStartHook(
```

Comment:
Could there be any downsides for not calling parsingDidStart/DidEnd in this code path (and not providing some alternate callback to take its place)?

Reply:
If someone has implemented critical `parsingDidStart` or `parsingDidEnd` functionality that didn't isolate its reach to that of the parsing phase, then they might find these callbacks are no longer invoked where they used to be.

Since we never documented or officially supported the `graphql-extensions` API and the request pipeline plugin hooks are new (though still undocumented. see #2008), I think the best we can do is stay true to the name of the methods and stand by the fact that parsing did not start or end; I don't think we should call `parsingDidStart` and `parsingDidEnd` if those behaviors did not actually happen.

Presumably, those who wanted behavior which occurred when execution started or ended would be using the `executionDidStart` or `executionDidEnd` hooks (and more generally, `requestDidStart`, `requestDidEnd`, `willSendResponse`, etc.).

I don't think there are any alternate callbacks that we can provide that will make this any easier, and I don't think such hooks would provide intrinsic value.